Version control multiple files

cmester0 opened the pull request:
Re-implementation of version control (#218).

franziskuskiefer commented on language/src/main.rs:
```diff
@@ -457,8 +452,22 @@ fn handle_crate<'tcx>(
+                    let file_temp_dir = original_file.join("_temp");
+                    let file_temp_dir = file_temp_dir.to_str().unwrap();
+            
+                    std::fs::create_dir_all(file_temp_dir.clone()).expect("Failed to crate dir");
```
```suggestion
                    std::fs::create_dir_all(file_temp_dir.clone()).expect("Failed to create dir");
```

franziskuskiefer commented on language/src/main.rs:
```diff
@@ -48,6 +48,7 @@ struct HacspecCallbacks {
     output_directory: Option<String>,
     output_type: Option<String>,
     target_directory: String,
+    version_control: Option<bool>
```
This is an odd type...
How about usig an enum with `Init` and `Update` values instead?

franziskuskiefer commented on language/src/main.rs:
```diff
@@ -773,6 +832,21 @@ fn main() -> Result<(), usize> {
         None => None,
     };
 
+    let init_file = match args.iter().position(|a| a == "--init") {
```
If the arguments are only `init` and `update`, it's not entirely clear what they refer to.
Also the other args use only one `-`.

Maybe something like `-vcs-init` could work?

cmester0 replied:
usually you use `-` for one letter arguments, while you use `--` for longer arguments. I guess we should update dir to be `--` in that case.

franziskuskiefer reviewed (changes requested):
Nice getting this back in. A couple comments.
Could you also add this to the help text in `util.rs`? (We should eventually move to clap again to make this easier...)

franziskuskiefer commented on language/src/main.rs:
```diff
@@ -744,7 +811,7 @@ fn main() -> Result<(), usize> {
 
     // Optionally get output directory.
-    let output_directory_index = args.iter().position(|a| a == "-dir");
+    let output_directory_index = args.iter().position(|a| a == "--dir");
```
This works too. But now you have to change all the calls in CI 🙂

franziskuskiefer commented on language/src/main.rs:
```diff
@@ -773,6 +840,21 @@ fn main() -> Result<(), usize> {
         None => None,
     };
 
+    let vc = match args.iter().position(|a| a == "--vc-init") {
```
Can you make an example somewhere (preferably with CI integration) that uses this and describe how to use it in the book? We can do this in a follow up if you prefer.

franziskuskiefer reviewed (changes requested):
lgtm now but you have to fix the `-dir` calls before we can merge.

cmester0 commented:
The tests work on my local computer, I am not sure why it is not allowed to copy the file to the `_vc` folder. I think, I had a similar issue, when I tried making tests for the backends.

franziskuskiefer commented:
> The tests work on my local computer, I am not sure why it is not allowed to copy the file to the `_vc` folder. I think, I had a similar issue, when I tried making tests for the backends.

It works for me as well locally but it looks like the directory is not created? Maybe some issue with the Rust or Linux version?

cmester0 commented:
@franziskuskiefer This should be ready for merge now, fixed the CI issues and added a test that VC can generate files. We should probably remove the generated files from the repository, or add them to some cleanup script, since the existance of old test files was why it worked locally.

franziskuskiefer commented on language/tests/fstar/QuestionMark.fst:
```diff
@@ -1,162 +0,0 @@
-module QuestionMark
```
If you mean removing these files, I agree. We should remove all generated files here from the repo and re-generate them on CI when they are needed.

franziskuskiefer reviewed (approved):
Thanks, lgtm.
Do you want to remove the files in this PR or do a separate one?
